db/range_del_aggregator.cc

```diff
@@ -365,29 +271,208 @@ Status RangeDelAggregator::AddTombstone(RangeTombstone tombstone) {
+        ++rep_iter;
         untermed_seq = kMaxSequenceNumber;
       } else {
         ++new_range_dels_iter;
-        ++tombstone_map_iter;
+        ++rep_iter;
         untermed_seq = kMaxSequenceNumber;
       }
     }
     while (new_range_dels_iter != new_range_dels.end()) {
-      tombstone_map.emplace(
-          new_range_dels_iter->start_key_,
-          RangeTombstone(Slice(), Slice(), new_range_dels_iter->seq_));
+      rep_.emplace(new_range_dels_iter->start_key_,
+                   RangeTombstone(Slice(), Slice(), new_range_dels_iter->seq_));
       ++new_range_dels_iter;
     }
+  }
+
+  size_t Size() const { return rep_.size() - 1; }
+
+  void InvalidatePosition() { iter_ = rep_.end(); }
+
+  typedef std::multimap<Slice, RangeTombstone, stl_wrappers::LessOfComparator>
+      Rep;
+
+  class Iterator : public RangeDelIterator {
+   public:
+    Iterator(const Rep& rep) : rep_(rep), iter_(rep.begin()) {}
+
+    bool Valid() const override {
+      return iter_ != rep_.end() && std::next(iter_) != rep_.end();
+    }
+
+    void Next() override {
+      do {
+        iter_++;
+      } while (Valid() && iter_->second.seq_ == 0);
+    }
+
+    RangeTombstone Tombstone() const override {
+      RangeTombstone tombstone;
+      tombstone.start_key_ = iter_->first;
+      tombstone.end_key_ = std::next(iter_)->first;
+      tombstone.seq_ = iter_->second.seq_;
+      return tombstone;
+    }
+
+   private:
+    const Rep& rep_;
+    Rep::const_iterator iter_;
+  };
+
+  std::unique_ptr<RangeDelIterator> NewIterator() {
+    return std::unique_ptr<RangeDelIterator>(new Iterator(this->rep_));
+  }
+
+ private:
+  Rep rep_;
+  Rep::iterator iter_;
+  const Comparator* ucmp_;
+};
+
+RangeDelAggregator::RangeDelAggregator(
+    const InternalKeyComparator& icmp,
+    const std::vector<SequenceNumber>& snapshots,
+    bool collapse_deletions /* = true */)
+    : upper_bound_(kMaxSequenceNumber),
+      icmp_(icmp),
+      collapse_deletions_(collapse_deletions) {
+  InitRep(snapshots);
+}
+
+RangeDelAggregator::RangeDelAggregator(const InternalKeyComparator& icmp,
+                                       SequenceNumber snapshot,
+                                       bool collapse_deletions /* = false */)
+    : upper_bound_(snapshot),
+      icmp_(icmp),
+      collapse_deletions_(collapse_deletions) {}
+
+void RangeDelAggregator::InitRep(const std::vector<SequenceNumber>& snapshots) {
+  assert(rep_ == nullptr);
+  rep_.reset(new Rep());
+  for (auto snapshot : snapshots) {
+    rep_->stripe_map_.emplace(snapshot, NewRangeDelMap());
+  }
+  // Data newer than any snapshot falls in this catch-all stripe
+  rep_->stripe_map_.emplace(kMaxSequenceNumber, NewRangeDelMap());
+  rep_->pinned_iters_mgr_.StartPinning();
+}
+
+std::unique_ptr<RangeDelMap> RangeDelAggregator::NewRangeDelMap() {
+  RangeDelMap* tombstone_map;
+  if (collapse_deletions_) {
+    tombstone_map = new CollapsedRangeDelMap(icmp_.user_comparator());
```

Comment:
Can't this be `return new CollapseRangeDelMap(...)`? Or does the `std::unique_ptr<>` not allow for that?

Reply:
Yeah, doesn't compile. Perhaps there's some C++ magic I don't know about, though.

Reply:
Yeah, I see that `unique_ptr` doesn't allow for that. Carry on.